### ISSUE TITLE
[OpenGL] PSSM shadow causes low resolution textures

### ISSUE THREAD

**Issue by Speciesx** [RigsOfRods/rigs-of-rods#640]
PSSM shadow on
![1](https://cloud.githubusercontent.com/assets/1196375/12538885/c553d086-c2e5-11e5-9100-3530758be3bc.jpg)

PSSM shadow off
![2](https://cloud.githubusercontent.com/assets/1196375/12538884/c553791a-c2e5-11e5-977c-1211a20b424a.jpg)

**Comment by Hiradur** [comment]
Doesn't happen with Nvidia GPU on Linux with proprietary drivers.

**Comment by Michael10055** [comment]
Doesn't happen here either.
![image](https://user-images.githubusercontent.com/11002490/28802930-0f7de5c0-7627-11e7-8e36-1dc4d5e713b2.png)

**Comment by Speciesx** [comment]
Still valid for me.
![ror_2017_08_01_18_03_07_809](https://user-images.githubusercontent.com/1196375/28834907-f70e2e38-76e3-11e7-9467-61904aac8cb0.jpg)

**Comment by only-a-ptr** [comment]
@Speciesx What's your OS/hardware? Post RoR.log.

**Comment by fidoj** [comment]
i seen this through testing, this is related to pssm layer drawing. ive caused this, ill see how to reproduce it

**Comment by Speciesx** [comment]
@only-a-ptr 
Win 10 Pro x64
i5 4670
Radeon RX 480

[RoR.log.txt](https://github.com/RigsOfRods/rigs-of-rods/files/1194350/RoR.log.txt)

**Comment by fidoj** [comment]
@Speciesx did you self compile or are you using prebuilt

**Comment by Speciesx** [comment]
@fidoj 
Compile myself.

**Comment by fidoj** [comment]
the error points to missing config files. specifically the shader, and this is where i seen it befor trying to change the shader, made it invalid and it will only display the min layer. maybe check your skeleton and dependencies to make sure its all up to date

**Comment by Michael10055** [comment]
Unfortunately I get the same issue now with my integrated Intel GPU. 
![image](https://user-images.githubusercontent.com/11002490/33246484-da4d9314-d2e2-11e7-992a-0d7bd0cf49f0.png)

A very interesting error does appear in the log though:
```
19:38:07:   ## Initializing Shadow Subsystem
19:38:07: OGRE EXCEPTION(7:InternalErrorException): Cannot load GL vertex program Ogre/ShadowExtrudeDirLightFiniteDebug.  Line 297:
Error, line 12: unknown identifier 'R1'.
Error, line 12: invalid destination register.
Error, line 13: unknown identifier 'R1'.
Error, line 13: unknown identifier 'R1'.
Error, line 13: invalid destination register.
Error, line 13: invalid source register.
Error, line 14: unknown identifier 'R1'.
Error, line 14: invalid source register.
 in Ogre/ShadowExtrudeDirLightFiniteDebug at D:\ror\dev\dependencies\src\Ogre\ogre_src_v1-9\RenderSystems\GL\src\OgreGLGpuProgram.cpp (line 251)
19:38:07: Gpu program Ogre/ShadowExtrudeDirLightFiniteDebug encountered an error during loading and is thus not supported.
```

**Comment by only-a-ptr** [comment]
@Michael10055 Intel GPUs are notorious troublemakers :worried: Can you try using DirectX?

The error is unusual but not strange - apparently the GL subsystem doesn't have a mechanism for rejecting unsupported features, it just rejects them as errors.

**Comment by AnotherFoxGuy** [comment]
This is probably caused by the CG shaders that are used for the pssm shadows

**Comment by Michael10055** [comment]
@only-a-ptr DirectX works, but I get the the flickering issue (#361)

**Comment by only-a-ptr** [comment]
Removing from v0.4.8 milestone - let's port to OGRE 1.10+ first.

**Comment by Ocawesome101** [comment]
four years later i still get this issue on two AMD GPUs (Radeon Vega 7 and RX 560) and I believe I've encountered it on an older NVidia GPU (GT M 330) and an Intel iGPU as well.  all of them on Linux, and all of them using OpenGL.

**Comment by tritonas00** [comment]
> four years later i still get this issue on two AMD GPUs (Radeon Vega 7 and RX 560) and I believe I've encountered it on an older NVidia GPU (GT M 330) and an Intel iGPU as well. all of them on Linux, and all of them using OpenGL.

Hi! We currently porting to Ogre 13 and remaking our shadows and lighting with RTSS.

**Comment by Ocawesome101** [comment]
excellent!  i look forward to that being completed.

**Comment by cryham** [comment]
Right. I have this too (basically darker terrain without textures) on Debian 12 with OpenGL,
in recent dev build from:
https://rigs-of-rods.itch.io/rigs-of-rods-dev
https://forum.rigsofrods.org/threads/ror-development-builds-for-windows-and-linux.696/
Which is still using Ogre 1.11 I see.

I noticed the branch ogre-13 here
https://github.com/RigsOfRods/rigs-of-rods/commits/ogre-13/
Seem done already? But not merged.
https://github.com/RigsOfRods/rigs-of-rods/pull/2858

With info in here, 5 months ago
https://github.com/RigsOfRods/rigs-of-rods/commit/f5c0183a11476fb327f3707863512100d3b82e2a
about Ogre 14, but with "crashes when loading terrain"

So I got about 2 questions:
- why isn't ogre-13 merged, so that dev build could be updated from ogre 1.11
- are you maybe moving to OgreNext or Ogre 2.x etc?

I'm not on discord, so if there was any info there, I'd miss it.
I think It'd be nice to see a post or something about dev plans / progress.

I already saw [this post](https://forum.rigsofrods.org/threads/checklist-for-porting-to-different-3d-engine-than-ogre.3510/) and [that post](https://forum.rigsofrods.org/threads/enter-the-codebase.121/post-20872) also [this](https://forum.rigsofrods.org/threads/checklist-for-porting-to-different-3d-engine-than-ogre.3510/).

**Comment by tritonas00** [comment]
@cryham 

We started the Ogre 13 branch to experiment with RTSS terrain, per pixel lighting, RTSS PSSM3 shadows and PBR. Although RTSS shadows are much much better, we experienced performance issues, mostly due to lighting. Then very limited time happened (two jobs) and i didn't investigate more. Now i have a bit more time, but shaders got my head :laughing:

RoR lacks dev power. There is one big boy (@ohlidalp) and he does mostly internal works, no graphics. I really do care about graphics, but i lack knowledge to finish suck a big task (my vision): Latest Ogre, RTSS (terrain, per pixel lighting, shadows), one nice sky shader, one nice water shader and one modern PBR based terrain. Maybe a vehicle also.

You certainly need to come in RoR discord, the main developer is more active there and i'm sure he can explain the situation much better than me, about everything.

PS: Very nice to see such an experienced programmer around here :+1:

**Comment by cryham** [comment]
Okay thanks for info.
I don't have much knowledge of recent Ogre 14 and almost none of RTSS.
I kind of skipped that when moving with SR3 to OgreNext (3.0), from an old Ogre 1.10 etc with shiny which was abandoned. I didn't yet restore all we had, still lots to do and fix. Yeah SR3 lacks devs too.

Right it's surely a good move to finish that ogre-13 branch and your vision.

Well OgreNext was a big task for me, took a year more or less, definitely advanced, I didn't expect so many changes, well we also have RTTs in editor, own terrain blendmap etc. But still, was surprising that even texture wrap wasn't easy and new materials etc.
So in case you wanted to someday try/move into OgreNext, I got a long [Ogre thread](https://forums.ogre3d.org/viewtopic.php?t=96576) with Q&A and my issues, then I wrote a summary with info in our [Developing.md](https://github.com/stuntrally/stuntrally3/blob/main/docs/Developing.md) well some of it SR3 specific. And BTW I also made that Ogre3 [terrain demo](https://github.com/cryham/ogre3ter-demo), I think it's a good starting point.

**Comment by cryham** [comment]
Also since Ogre 13.5 there is [automatic instancing](https://github.com/OGRECave/ogre/releases/tag/v13.5.0).
I'd say worth trying it and maybe even paged-geom could be replaced. In SR3 I do it this way, all vegetation trees etc are just Items (old Entities), this requires good LODs for those meshes to lower total triangle count but in OgreNext even LODs are in same batch which is great also terrain is quite good there too.

Anyways I've built Ogre-13 branch and indeed terrain crashes, only "Simple test terrain" maps load and some cars have single color materials. Not sure if that's normal.
Here are some errors from log:
```
Error: RTSS: failed to create GpuPrograms for Pass 0 of 'Sandrail-LS-driver@Actor_0'
RTSS: using 3bb7ceab2f420016c17ece58285f718b_VS for Pass 0 of 'Sandrail-LS-driver2@Actor_0'
RTSS: using 1a3e0a34fe60ea24c27b8a786ab73296_FS for Pass 0 of 'Sandrail-LS-driver2@Actor_0'
Error: RTSS: failed to create GpuPrograms for Pass 0 of 'Sandrail-LS-driver2@Actor_0'
Program '3775bfdcd175ad3fe1ed0b9c9d5eebc0_FS' is not supported: '3775bfdcd175ad3fe1ed0b9c9d5eebc0_FS' 0:14(2): error: image/sampler variables may only be declared as function parameters or uniform-qualified global variables
0:14(2): error: opaque variables must be declared uniform
0:15(2): error: image/sampler variables may only be declared as function parameters or uniform-qualified global variables
0:15(2): error: opaque variables must be declared uniform
0:47(20): warning: `gTextureSampler0' used uninitialized
0:48(20): warning: `gTextureSampler1' used uninitialized
0:49(20): warning: `gTextureSampler1' used uninitialized
```
Ok actually I can load terrains in Release, it's white but didn't crash.
Had to first comment this out isNaN wasn't there.
```
    // r = engine->RegisterObjectMethod("quaternion", "bool isNaN() const", asMETHOD(Quaternion,isNaN), asCALL_THISCALL);
    // ROR_ASSERT( r >= 0 );
```

**Comment by cryham** [comment]
Actually errors start even earlier like so:
```
21:18:22: Warning: ScriptCompiler - deprecated symbol in texture_manager.material(16): 'cubic_texture ..'. Use 'texture .. cubic' instead.
21:18:22: Parsing script managed_submesh.material
21:18:22: Error: ScriptCompiler - reference to a non existing object in managed_mats.material(1): shadows.material
21:18:22: Error: ScriptCompiler - base object not found in managed_mats.material(58): Shadows/managed_transparent/base_receiver
21:18:22: Error: ScriptCompiler - base object not found in managed_mats.material(76): Shadows/managed_transparent/base_receiver
```
and for terrain:
```
21:18:26: Terrain created; size=1025 minBatch=17 maxBatch=65 treeDepth=5 lodLevels=7 leafLods=3
21:18:27: Program '2f2941af9412051ecb0d5e590f99c04b_FS' is not supported: '2f2941af9412051ecb0d5e590f99c04b_FS' 0:16(2): error: image/sampler variables may only be declared as function parameters or uniform-qualified global variables
0:16(2): error: opaque variables must be declared uniform
0:17(2): error: image/sampler variables may only be declared as function parameters or uniform-qualified global variables
0:17(2): error: opaque variables must be declared uniform
0:18(2): error: image/sampler variables may only be declared as function parameters or uniform-qualified global variables
0:18(2): error: opaque variables must be declared uniform
0:19(2): error: image/sampler variables may only be declared as function parameters or uniform-qualified global variables
0:19(2): error: opaque variables must be declared uniform
0:20(2): error: image/sampler variables may only be declared as function parameters or uniform-qualified global variables
0:20(2): error: opaque variables must be declared uniform
0:21(2): error: image/sampler variables may only be declared as function parameters or uniform-qualified global variables
0:21(2): error: opaque variables must be declared uniform
0:22(2): error: image/sampler variables may only be declared as function parameters or uniform-qualified global variables
0:22(2): error: opaque variables must be declared uniform
0:53(18): warning: `globalNormal0' used uninitialized
0:56(25): warning: `blendTex1' used uninitialized
0:57(2): error: no function with name 'SGX_CalculateTerrainTBN'
0:60(52): warning: `difftex2' used uninitialized
0:61(59): warning: `difftex3' used uninitialized
0:62(59): warning: `difftex4' used uninitialized
0:63(59): warning: `difftex5' used uninitialized
0:64(59): warning: `difftex6' used uninitialized
0:67(2): error: no function with name 'evaluateLight'
0:68(2): error: no function with name 'evaluateLight'
0:69(2): error: no function with name 'evaluateLight'
```
There is no source for this, it is generated.


Okay wow I actually got visible terrain.
Freaking dark, but I see its textures are blending okay.

![09_21-33-24`](https://github.com/RigsOfRods/rigs-of-rods/assets/720641/788c91c2-30f9-43ec-9848-e63858a54218)

What I did was getting latest Ogre from git, and then:
copying/packing all from Ogre's `Media/Main/*` into RoR's `build/bin/resources/OgreCore.zip`,
and same from Ogre's `Media/RTShaderLib` into RoR's `build/bin/resources/rtshader.zip`

**Comment by cryham** [comment]
So, where is or was `shadows.material` (not found now)
I see few places in `managed_mats*.material` with
`import * from "shadows.material"`
which are failing now on ogre-13 branch.
What was there and will it be even needed in RTSS? I'm guessing not.

**Comment by tritonas00** [comment]
> So, where is or was shadows.material (not found now)

https://github.com/RigsOfRods/rigs-of-rods/blob/master/resources/managed_materials/shadows/pssm/on/shadows.material

> which are failing now on ogre-13 branch.
What was there and will it be even needed in RTSS? I'm guessing not.
Did it only have Shadows/managed_transparent/base_receiver?

Nope. Ogre 13 branch shadows don't use any of those. It's code only, using RTSS PSSM3 and trouble free actually, even on linux.

**Comment by cryham** [comment]
Okay.
Well I'm getting somewhere. Also fixing some .material warnings and .fontdef.
I should open a PR ASAP.
![10_11-46-35``](https://github.com/RigsOfRods/rigs-of-rods/assets/720641/7cddd79c-8f9f-4522-a8d6-f569f6ca5a53)

**Comment by tritonas00** [comment]
I think you don't have cg installed btw.

**Comment by cryham** [comment]
Well yes, I don't. Is CG needed still? It's old and bad.
I thought we'd remove that BTW of RTSS and ogre 14.
I'm on Debian 12, with amd-gpu.

**Comment by cryham** [comment]
There was also this https://github.com/RigsOfRods/rigs-of-rods/issues/1412
and it's closed.
And yeah lots of tracks have some *.cg files making meshes invisible.
![cg1](https://github.com/RigsOfRods/rigs-of-rods/assets/720641/c27eb725-0322-4727-aad7-23a8f73b6b8e)
![cg2](https://github.com/RigsOfRods/rigs-of-rods/assets/720641/2529d9ef-03ef-4373-a6a7-eb0073e0440d)
![cg3](https://github.com/RigsOfRods/rigs-of-rods/assets/720641/0db22a47-f3bd-4874-84fe-e93da60238fe)

**Comment by tritonas00** [comment]
> Well yes, I don't. Is CG needed still? It's old and bad.

RoR massively uses it, sadly.

**Comment by cryham** [comment]
Yes. This would require updating many tracks and vehicles. IDK if that's okay for anyone to do (upload a fixed version w/o *.cg and changed materials), or is it something you'd need author's permission first (i.e. won't be possible)?

**Comment by cryham** [comment]
Ah I see now there is also ogre-14, I'm gonna check that first and hold on with my PR.

**Comment by CuriousMike56** [comment]
> Yes. This would require updating many tracks and vehicles. IDK if that's okay for anyone to do (upload a fixed version w/o *.cg and changed materials), or is it something you'd need author's permission first (i.e. won't be possible)?

As someone who's spent countless hours improving and re-uploading hundreds of mods on the Repository, fixing each mod isn't feasible. We need a way to automatically convert Cg materials. 
AFAIK Cg is the only dependency we use that's closed source (Why Nvidia never open sourced it is beyond me). Not only does it prevent us from porting to newer Ogre. it also prevents porting the game to other platforms/architectures.

**Comment by cryham** [comment]
Ok I understand. IDK maybe there is some code or librarly that could convert. If not then new RoR code to read cg files? Sounds complicated.
IIRC we had a discussion about Cg in SR team (when there were other devs) back in 2012 I guess, and decided to drop it early. But at that time scrawl also worte 1st material generator and then shiny which were cool to use back then.

Anyways I tried that ogre-14 branch, seems similar to ogre-13? I had a couple of issues needed to fix build.
Now somehow I have white terrain. I'll try to compare.
And I also built paged-geom which now sometimes crashes too on track load.
I also get a crash after a moment inside RoR when accesing Repository (seems for images?).

I think I'll instead, get into trying that "export SR track to RoR" thing in SR3 editor, should be easier to do.

**Comment by cryham** [comment]
Ok I did a PR https://github.com/RigsOfRods/rigs-of-rods/pull/3124
To not loose any of that. Could be helpful. Is quite WIP.
